<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.361">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>Can sparse Bayesian independent component analysis describe gene regulation?</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for citations */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging-indent div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}</style>


<script src="groundwork_files/libs/clipboard/clipboard.min.js"></script>
<script src="groundwork_files/libs/quarto-html/quarto.js"></script>
<script src="groundwork_files/libs/quarto-html/popper.min.js"></script>
<script src="groundwork_files/libs/quarto-html/tippy.umd.min.js"></script>
<script src="groundwork_files/libs/quarto-html/anchor.min.js"></script>
<link href="groundwork_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="groundwork_files/libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="groundwork_files/libs/bootstrap/bootstrap.min.js"></script>
<link href="groundwork_files/libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="groundwork_files/libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">

  <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

</head>

<body>

<div id="quarto-content" class="page-columns page-rows-contents page-layout-article">
<div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
  <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#background" id="toc-background" class="nav-link active" data-scroll-target="#background">Background</a>
  <ul class="collapse">
  <li><a href="#introduction" id="toc-introduction" class="nav-link" data-scroll-target="#introduction">Introduction</a></li>
  <li><a href="#imodulons" id="toc-imodulons" class="nav-link" data-scroll-target="#imodulons">Imodulons</a></li>
  <li><a href="#independent-component-analysis" id="toc-independent-component-analysis" class="nav-link" data-scroll-target="#independent-component-analysis">Independent Component Analysis</a></li>
  <li><a href="#transcriptomics-data" id="toc-transcriptomics-data" class="nav-link" data-scroll-target="#transcriptomics-data">Transcriptomics data</a></li>
  </ul></li>
  <li><a href="#our-project" id="toc-our-project" class="nav-link" data-scroll-target="#our-project">Our project</a>
  <ul class="collapse">
  <li><a href="#why-we-want-to-attempt-a-bayesian-imodulon-analysis" id="toc-why-we-want-to-attempt-a-bayesian-imodulon-analysis" class="nav-link" data-scroll-target="#why-we-want-to-attempt-a-bayesian-imodulon-analysis">Why we want to attempt a Bayesian Imodulon analysis</a></li>
  <li><a href="#what-we-would-like-to-achieve" id="toc-what-we-would-like-to-achieve" class="nav-link" data-scroll-target="#what-we-would-like-to-achieve">What we would like to achieve</a></li>
  <li><a href="#model" id="toc-model" class="nav-link" data-scroll-target="#model">Model</a></li>
  </ul></li>
  <li><a href="#where-we-need-help" id="toc-where-we-need-help" class="nav-link" data-scroll-target="#where-we-need-help">Where we need help</a></li>
  </ul>
<div class="quarto-alternate-formats"><h2>Other Formats</h2><ul><li><a href="groundwork.docx"><i class="bi bi-file-word"></i>MS Word</a></li></ul></div></nav>
</div>
<main class="content page-columns page-full" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Can sparse Bayesian independent component analysis describe gene regulation?</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  

</header>

<p>We think we could use a sparse Bayesian independent component analysis model to describe gene regulation in <em>E. coli</em> and other organisms for which there is RNA sequencing data. This document explains why we think this, how we would like do it and where we need help.</p>
<section id="background" class="level1 page-columns page-full">
<h1>Background</h1>
<section id="introduction" class="level2">
<h2 class="anchored" data-anchor-id="introduction">Introduction</h2>
<p>Living cells respond to environmental changes by up- or down-regulating their genes, thereby changing the proportion in which the cell produces proteins, which in turn alters the cell’s behaviour. For example, a cell that moves into a toxic environment might alter its gene expression so as to increase production of a protein that exports the toxin or to decrease toxin-importing proteins.</p>
<p>By analysing RNA transcripts it is possible to measure gene expression and see how it varies across conditions. There is now so much transcriptomics data available that it plauibly contains sufficient information to learn how some organisms orchestrate their gene regulation. <span class="citation" data-cites="tanIndependentComponentAnalysis2020">Tan et al. (<a href="#ref-tanIndependentComponentAnalysis2020" role="doc-biblioref">2020</a>)</span> achieved this goal using a method based on independent component analysis or ICA, leading to the development of the concept of an Imodulon.</p>
</section>
<section id="imodulons" class="level2 page-columns page-full">
<h2 class="anchored" data-anchor-id="imodulons">Imodulons</h2>
<p>An Imodulon is a hypothetical latent allocation of weights to a subset of genes derived from the results of an analysis involving ICA, that is taken to represent a way that a cell can regulate its genes in response to changing conditions. For example, suppose a certain Imodulon I1 regulates just two genes G1 and G2 with respective weights 0.5 and -0.5. When I1 is activated, G1 will be up-regulated and G2 will be down-regulated by the same amount. In contrast, when I1 is deactivated, the opposite regulation will occur. It is typically presumed that there are far fewer Imodulons than genes, that an Imodulon will substantially up- or down-regulate the genes that it affects, and that most Imodulons will affect a relatively small number of genes.</p>
<div class="page-columns page-full"><p>It is plausible that Imodulons roughly describe how gene regulation works because of the known existence of transcription units, transcription factors and regulons. Transcription units are sets of genes that share an RNA binding site and can therefore only be regulated together. Transcription factors are proteins that activate or deactivate particular transcription units. Regulons are sets of genes that are regulated by exactl the same transcription factors. Since there are known to be many of all these things, it is likely that a latent representation like Imodulons is roughly correct. In particular, based on the <a href="https://regulondb.ccg.unam.mx/index.jsp">regulonDB</a> it is clear that most transcription factors affect a small number of genes and that there are fewer transcription factors than genes.<a href="#fn1" class="footnote-ref" id="fnref1" role="doc-noteref"><sup>1</sup></a></p><div class="no-row-height column-margin column-container"><li id="fn1"><p><sup>1</sup>&nbsp;Note that Imodulons are not the same as transcription factors!</p></li></div></div>
<p>Whereas previous analyses have attempted to fit ICA models using an approach based on optimisation, we would like to fit a Bayesian statistical model that implements ICA. We would like our statistical model to include an explicit representation of the assumptions about Imodulon sparsity outlined above.</p>
</section>
<section id="independent-component-analysis" class="level2 page-columns page-full">
<h2 class="anchored" data-anchor-id="independent-component-analysis">Independent Component Analysis</h2>
<div class="page-columns page-full"><p>Independent component analysis assumes that the numbers comprising an <span class="math inline">\(I\times J\)</span> matrix of observations <span class="math inline">\(X\)</span> are generated by taking weighted sums of a known number <span class="math inline">\(K &lt; J\)</span> independent component vectors, as shown below: <a href="#fn2" class="footnote-ref" id="fnref2" role="doc-noteref"><sup>2</sup></a></p><div class="no-row-height column-margin column-container"><li id="fn2"><p><sup>2</sup>&nbsp;Term names chosen for consistency with the Imodulon papers</p></li></div></div>
<p><span id="eq-ica-long"><span class="math display">\[
\begin{equation}
  x_{ij}=\sum_{k=1}^{K} m_{ik}a_{kj}
\end{equation}
\tag{1}\]</span></span></p>
<p>or in matrix notation</p>
<p><span id="eq-ica"><span class="math display">\[
\begin{equation}
X = MA
\end{equation}
\tag{2}\]</span></span></p>
<p>It is assumed that the columns of the matrix <span class="math inline">\(A\)</span> are column-wise probabilistically independent, so that the probability of the <span class="math inline">\(j\)</span>th column of <span class="math inline">\(A\)</span> is the product of the <span class="math inline">\(K\)</span> marginal probabilities, i.e.&nbsp;<span class="math inline">\(p(a_{:j}) = \prod_{k=1}^I p(a_{kj})\)</span>. Secondly, it is also assumed that the rows of <span class="math inline">\(A\)</span> have non-Gaussian marginal distributions. See <span class="citation" data-cites="hyvarinenIndependentComponentAnalysis2000">Hyvärinen and Oja (<a href="#ref-hyvarinenIndependentComponentAnalysis2000" role="doc-biblioref">2000</a>)</span> for a discussion of an optimisation based approach to Independent component analysis and <span class="citation" data-cites="robertsBayesianIndependentComponent2005">Roberts and Choudrey (<a href="#ref-robertsBayesianIndependentComponent2005" role="doc-biblioref">2005</a>)</span> for a discussion of Bayesian independent component analysis.</p>
<p>In the canonical application of ICA each row of <span class="math inline">\(X\)</span> represents a time course of signals from a receiver detecting input from <span class="math inline">\(k\)</span> sources; each row of <span class="math inline">\(A\)</span> represents the time course of signals from a source; each column of <span class="math inline">\(M\)</span> represents how a source mixes between receivers. In the context of a transcriptomics data analysis the observation units are genes rather than receivers, the observation rows represent separate experiments rather than time courses and the columns of the matrix <span class="math inline">\(M\)</span> represents proto-Imodulons, i.e. mixing weights for each gene for each proto-Imodulon. A separate downstream analysis is required in order to sparsify the results by removing genes from Imodulons and discarding some candidate Imodulons.</p>
<p>The Python library Scikit learn provides access to <a href="https://scikit-learn.org/stable/modules/generated/sklearn.decomposition.FastICA.html">an implementation of ICA</a> based on minimsation of mutual information, as outlined in <span class="citation" data-cites="hyvarinenIndependentComponentAnalysis2000">Hyvärinen and Oja (<a href="#ref-hyvarinenIndependentComponentAnalysis2000" role="doc-biblioref">2000</a>)</span>.</p>
</section>
<section id="transcriptomics-data" class="level2">
<h2 class="anchored" data-anchor-id="transcriptomics-data">Transcriptomics data</h2>
<p>The <a href="https://github.com/SBRG/precise-db">precise-db</a> provides a large collection of RNA sequencing data that can be used to create suitable input for independent component analysis.</p>
<p>Like with other RNA sequencing data, each gene expression measurement in precise-db starts as a count of the number of times an mRNA fragment that maps to the gene was detected in the experiment fragments per transcript. Unfortunately such counts are not comparable between genes within a sample because they are sensitive to the size of the mRNA fragment that encodes the gene (a big gene will tend to be counted more often than an equally expressed small gene). Raw counts are also not comparable between samples because of potential variations in sequence depth (i.e.&nbsp;the total number of measurements from the sample) and the mRNA-fragment-to-gene map used. Consequently the usual practice is to transform raw counts by first normalising based on the gene size and then ‘proportionising’, ending with a relative unit called transcripts per million or TPM:</p>
<p><span id="eq-tpm"><span class="math display">\[
\begin{equation}
TPM(g) = \frac{count(g)/size(g)}{\sum_{i\in sample}count(i)/size(i)}
\end{equation}
\tag{3}\]</span></span></p>
<p>See <span class="citation" data-cites="zhaoMisuseRPKMTPM2020">Zhao, Ye, and Stanton (<a href="#ref-zhaoMisuseRPKMTPM2020" role="doc-biblioref">2020</a>)</span> for discussion of the transcripts per million normalisation and for references to more papers about RNA sequencing experiments.</p>
<p>For use in imodulon analysis further transformations are performed. First the transcripts per million are put on log scale, then the log-transcripts per million of a reference condition are subtracted. Finally the data are whitened using scikit-learn’s <code>arbitrary-variance</code> option.</p>
</section>
</section>
<section id="our-project" class="level1">
<h1>Our project</h1>
<section id="why-we-want-to-attempt-a-bayesian-imodulon-analysis" class="level2">
<h2 class="anchored" data-anchor-id="why-we-want-to-attempt-a-bayesian-imodulon-analysis">Why we want to attempt a Bayesian Imodulon analysis</h2>
<p>There are several general reasons to prefer Bayesian ICA models to models that use a maximum likelihood approach, including automatic relevance detection, the potential to take into account quantitative non-measurement information through a prior model, and the avoidance of pathological model behaviour due to bad or incomplete observations. These general reasons are discussed in <span class="citation" data-cites="robertsBayesianIndependentComponent2005">Roberts and Choudrey (<a href="#ref-robertsBayesianIndependentComponent2005" role="doc-biblioref">2005</a>)</span>.</p>
<p>These advantages are particularly pertinent for the application of ICA to attempting to infer Imodulons from RNA sequencing data.</p>
<p>First, as described in <span class="citation" data-cites="robertsBayesianIndependentComponent2005">Roberts and Choudrey (<a href="#ref-robertsBayesianIndependentComponent2005" role="doc-biblioref">2005</a>)</span>, a Bayesian ICA model can potentially use automatic relevance detection to find an appropriate number of Imodulons to postulate. Whereas in the current framework the appropriate number of Imodulons is determined using a procedure that is separate from the main inference and motivated by computational and algorithmic considerations—primarily whether the algorithm successfully converges—rather than substantive statistical ones, relevance detection based on hyperparameters in the context of a Bayesian model is well-motivated and does not require downstream processing.</p>
<p>Second, there is substantial non-experimental information about Imodulons that should provide important context to RNA sequencing measurements. In particular, as discussed above investigation of regulons and transcription factors suggests that an Imodulon should typically affect a relatively small number of genes. Maximum-likelihood based Imodulon analysis uses another non-statistically-motivated downstream procedure to impose this sparsity after fitting an ICA model. We would like to represent this information in a Bayesian ICA model by using a sparsity-inducing prior for the columns of the mixing matrix <span class="math inline">\(M\)</span>.</p>
<p>Another potential source of non-experimental information is the research into regulons and transcription factors itself. Many genes, particularly those in a species’s ‘core’ genome that are common to almost all strains, are known in advance to share specific regulators, and this information can be taken into account in a Bayesian framework through informative priors on specific elements of the mixing matrix <span class="math inline">\(M\)</span>.</p>
<p>Finally, the robustness gained by using a Bayesian statistical analysis is likely to be helpful when attempting to infer Imodulons for species with fewer RNA sequencing experiments.</p>
</section>
<section id="what-we-would-like-to-achieve" class="level2">
<h2 class="anchored" data-anchor-id="what-we-would-like-to-achieve">What we would like to achieve</h2>
<p>Ultimately we would like to create a sparse Bayesian ICA model that can reproduce the analysis in <span class="citation" data-cites="tanIndependentComponentAnalysis2020">Tan et al. (<a href="#ref-tanIndependentComponentAnalysis2020" role="doc-biblioref">2020</a>)</span> of the precise-db dataset. This analysis inferred 92 Imodulons from RNA sequencing measurements of 4386 genes in 278 conditions. We would then like to use the same approach to analyse RNA sequencing data from organisms with fewer measurements and to augment our model with informative priors based on information about regulons.</p>
<p>As a proof of concept we would like to generate and then fit a smaller artificial dataset with uninformative but still sparsity-inducing priors, then fit the same model to a subset of the precise-db dataset.</p>
</section>
<section id="model" class="level2">
<h2 class="anchored" data-anchor-id="model">Model</h2>
<p>In contrast to the model in <span class="citation" data-cites="hyvarinenIndependentComponentAnalysis2000">Hyvärinen and Oja (<a href="#ref-hyvarinenIndependentComponentAnalysis2000" role="doc-biblioref">2000</a>)</span> and previous Imodulon analyses we propose to include measurement noise in our statistical model. Although it would perhaps be preferable to model the whole data generation process up to the production of untransformed raw counts, for the sake of simplicity and easier comparison with previous approaches we will use the following simple linear regression model for transformed RNA sequencing data, assuming known measurement error <span class="math inline">\(\sigma\)</span>:</p>
<p><span id="eq-noise-likelihood"><span class="math display">\[
\begin{equation*}
y \sim N(\hat{y}, \sigma)
\end{equation*}
\tag{4}\]</span></span></p>
<p>In order to ensure that our model implements ICA we will use a column-wise independent and row-wise non-Gaussian prior distribution for the source strength matrix <span class="math inline">\(A\)</span>:</p>
<p><span id="eq-source-model"><span class="math display">\[
\begin{equation*}
a_kj \sim T4(0,1)
\end{equation*}
\tag{5}\]</span></span></p>
<p>In <a href="#eq-source-model">Equation&nbsp;5</a> <span class="math inline">\(T4\)</span> refers to the student-t distribution with 4 degrees of freedom. The use of a unit scale for each row is beause ICA models are identified only up to a change of scale</p>
<p>We will use independent regularised horseshoe priors for the columns of the mixing matrix <span class="math inline">\(M\)</span>:</p>
<p><span id="eq-mixing-model"><span class="math display">\[
\begin{align*}
m_{ik} &amp;\sim N^+(0, \tau_k\tilde{\lambda}_{ik}) \\
\tilde{\lambda}_{ik})^2 &amp;= \frac{c_k^2\lambda_{ik}^2}{c_k^2 + \tau_k^2\lambda_{ik}^2} \\
\lambda_{ik} &amp;\sim C^+(0, 1) \\
c_k^2 &amp;\sim \text{inverse gamma}(\alpha, \beta) \\
\tau_k &amp;\sim C^+(0, \tau_{k0})
\end{align*}
\tag{6}\]</span></span></p>
<p>In <a href="#eq-mixing-model">Equation&nbsp;6</a> <span class="math inline">\(C^+\)</span> and <span class="math inline">\(N^+\)</span> refer respectively to Cauchy and Normal distributions with support only on the positive real line, and the terms <span class="math inline">\(alpha\)</span>, <span class="math inline">\(\beta\)</span> and <span class="math inline">\(tau_{k0}\)</span> are informative priors. Note that the mixing matrix prior is constrained to have support only on the non-negative real line. This is done, following <span class="citation" data-cites="robertsBayesianIndependentComponent2005">Roberts and Choudrey (<a href="#ref-robertsBayesianIndependentComponent2005" role="doc-biblioref">2005</a>)</span>, to ensure sign consistency, so that a positive relative change in an Imodulon’s strength in a certain condition will always correspond to up-regulation of genes affected by that Imodulon.</p>
</section>
</section>
<section id="where-we-need-help" class="level1">
<h1>Where we need help</h1>
<p>We need to address a few statistical programming issues as we have not yet managed to implement our target model:</p>
<ul>
<li><p>Is it feasible to fit our target model and dataset with Stan, or should we try a different framework (or give up, try a simpler model etc)? What would be good potential options for simpler models or alternative frameworks?</p></li>
<li><p>What is the right way to implement the regularised horseshoe prior for this case?</p></li>
</ul>

</section>


<div id="quarto-appendix" class="default"><section class="quarto-appendix-contents" role="doc-bibliography"><h2 class="anchored quarto-appendix-heading">References</h2><div id="refs" class="references csl-bib-body hanging-indent" role="list">
<div id="ref-hyvarinenIndependentComponentAnalysis2000" class="csl-entry" role="listitem">
Hyvärinen, A., and E. Oja. 2000. <span>“Independent Component Analysis: Algorithms and Applications.”</span> <em>Neural Networks</em> 13 (4): 411–30. <a href="https://doi.org/10.1016/S0893-6080(00)00026-5">https://doi.org/10.1016/S0893-6080(00)00026-5</a>.
</div>
<div id="ref-robertsBayesianIndependentComponent2005" class="csl-entry" role="listitem">
Roberts, Stephen, and Rizwan Choudrey. 2005. <span>“Bayesian <span>Independent Component Analysis</span> with <span>Prior Constraints</span>: <span>An Application</span> in <span>Biosignal Analysis</span>.”</span> In <em>Deterministic and <span>Statistical Methods</span> in <span>Machine Learning</span></em>, edited by Joab Winkler, Mahesan Niranjan, and Neil Lawrence, 159–79. Lecture <span>Notes</span> in <span>Computer Science</span>. <span>Berlin, Heidelberg</span>: <span>Springer</span>. <a href="https://doi.org/10.1007/11559887_10">https://doi.org/10.1007/11559887_10</a>.
</div>
<div id="ref-tanIndependentComponentAnalysis2020" class="csl-entry" role="listitem">
Tan, Justin, Anand V. Sastry, Karoline S. Fremming, Sara P. Bjørn, Alexandra Hoffmeyer, Sangwoo Seo, Bjørn G. Voldborg, and Bernhard O. Palsson. 2020. <span>“Independent Component Analysis of <span>E</span>. Coli’s Transcriptome Reveals the Cellular Processes That Respond to Heterologous Gene Expression.”</span> <em>Metabolic Engineering</em> 61 (September): 360–68. <a href="https://doi.org/10.1016/j.ymben.2020.07.002">https://doi.org/10.1016/j.ymben.2020.07.002</a>.
</div>
<div id="ref-zhaoMisuseRPKMTPM2020" class="csl-entry" role="listitem">
Zhao, Shanrong, Zhan Ye, and Robert Stanton. 2020. <span>“Misuse of <span>RPKM</span> or <span>TPM</span> Normalization When Comparing Across Samples and Sequencing Protocols.”</span> <em>RNA</em> 26 (8): 903–9. <a href="https://doi.org/10.1261/rna.074922.120">https://doi.org/10.1261/rna.074922.120</a>.
</div>
</div></section></div></main>
<!-- /main column -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->



</body></html>